src/java.desktop/share/classes/javax/swing/plaf/metal/MetalBorders.java

```diff
@@ -246,115 +248,104 @@ public static class InternalFrameBorder extends AbstractBorder implements UIReso
          */
         public InternalFrameBorder() {}
 
-        /**
-         * Rounds a double to the nearest integer. It rounds 0.5 down,
-         * for example 1.5 is rounded to 1.0.
-         *
-         * @param d number to be rounded
-         * @return the rounded value
-         */
-        private static int roundHalfDown(double d) {
-            double decP = (Math.ceil(d) - d);
-            return (int)((decP == 0.5) ?  Math.floor(d) :  Math.round(d));
-        }
-
-
         public void paintBorder(Component c, Graphics g, int x, int y,
                                 int w, int h) {
-            Color background;
-            Color highlight;
-            Color shadow;
-
-            if (c instanceof JInternalFrame && ((JInternalFrame)c).isSelected()) {
-                background = MetalLookAndFeel.getPrimaryControlDarkShadow();
-                highlight = MetalLookAndFeel.getPrimaryControlShadow();
-                shadow = MetalLookAndFeel.getPrimaryControlInfo();
-            } else {
-                background = MetalLookAndFeel.getControlDarkShadow();
-                highlight = MetalLookAndFeel.getControlShadow();
-                shadow = MetalLookAndFeel.getControlInfo();
-            }
-
-            Graphics2D g2d = (Graphics2D) g;
-            AffineTransform at = g2d.getTransform();
-            Stroke oldStk = g2d.getStroke();
-            Color oldColor = g2d.getColor();
-            int stkWidth = 1;
-
-            // if m01 or m10 is non-zero, then there is a rotation or shear
-            // skip resetting the transform
-            boolean resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0));
-
-            int xtranslation;
-            int ytranslation;
-            int width;
-            int height;
-
-            if (resetTransform) {
-                g2d.setTransform(new AffineTransform());
-                stkWidth = roundHalfDown(Math.min(at.getScaleX(), at.getScaleY()));
-
-                double xx = at.getScaleX() * x + at.getTranslateX();
-                double yy = at.getScaleY() * y + at.getTranslateY();
-                xtranslation = roundHalfDown(xx);
-                ytranslation = roundHalfDown(yy);
-                width = roundHalfDown(at.getScaleX() * w + xx) - xtranslation;
-                height = roundHalfDown(at.getScaleY() * h + yy) - ytranslation;
-            } else {
-                width = w;
-                height = h;
-                xtranslation = x;
-                ytranslation = y;
-            }
-            g2d.translate(xtranslation, ytranslation);
-
-            // scaled border
-            int thickness = (int) Math.ceil(4 * at.getScaleX());
-
-            g.setColor(background);
-            // Draw the bulk of the border
-            for (int i = 0; i <= thickness; i++) {
-                g.drawRect(i, i, width - (i * 2), height - (i * 2));
-            }
+            if (g instanceof Graphics2D) {
```

Comment:
I think it's not the right approach. You should separate the parts of code into those which require `Graphics2D` (resetting transform, stroke width) and those which don't, so that the border is still painted even when `g` is not an instance of `Graphics2D`.

Take a look at how @alisenchung handled it for `EtchedBorder`:
https://github.com/openjdk/jdk/blob/b42c1ad1086a5c3f579e27380d23f67f8cebb437/src/java.desktop/share/classes/javax/swing/border/EtchedBorder.java#L159-L171

The border is still rendered even if the passed in `Graphics` is not `Graphics2D`.

Reply:
@aivanov-jdk Thanks for reviewing. Will check and make the necessary code changes.